src/main/java/bio/overture/ego/security/OAuth2ClientResources.java

```diff
@@ -24,10 +26,20 @@ public String getRedirectUri(AccessTokenRequest request) {
             val uri = new URI(request.getCurrentUri());
             val attr = (ServletRequestAttributes) RequestContextHolder.currentRequestAttributes();
             val session = attr.getRequest().getSession(true);
-            val pattern = Pattern.compile("client_id=([^&]+)?");
-            val matcher = pattern.matcher(uri.getQuery());
-            if (matcher.find()) {
-              session.setAttribute("ego_client_id", matcher.group(1));
+
+            val uriComponents = UriComponentsBuilder.fromUri(uri).build();
+            val queryParams = uriComponents.getQueryParams();
+
+            val clientId = queryParams.getFirst("client_id");
+            if (clientId != null && !clientId.isEmpty()) {
+              session.setAttribute("ego_client_id", clientId);
```

Comment:
since this and the redirect uri are extracted from the url, do we really need to store them in a session instead of extracting them again before redirect assuming we have access to the original request object ? im a bit ignorant on the redirection dance, so asking more for my information

Reply:
I believe its being used this way because the implementation isn't using the state parameter from the oauth2 spec correctly.

Reply:
To be explicit, instead of storing this in the session, we would put the clientid and redirect into a JSON object that we also make as a nonce and we base64 encode and put as the state.

Reply:
Basically have something like this:
```java
restTemplate.getOAuth2ClientContext().setPreservedState(ourStateInsteadOfSession);
```
in this class: https://github.com/overture-stack/ego/blob/develop/src/main/java/bio/overture/ego/security/OAuth2ClientResources.java